alert-combiner-py/src/agent.py

```diff
@@ -0,0 +1,157 @@
+    ALERTED_ADDRESSES = []
+
+    global FINDINGS_CACHE
+    FINDINGS_CACHE = []
+
+    global MUTEX
+    MUTEX = False
+
+
+def is_contract(w3, address) -> bool:
+    """
+    this function determines whether address is a contract
+    :return: is_contract: bool
+    """
+    if address is None:
+        return True
+    code = w3.eth.get_code(Web3.toChecksumAddress(address))
+    return code != HexBytes('0x')
+
+
+def detect_attack(w3, forta_explorer, block_event: forta_agent.block_event.BlockEvent):
+    """
+    this function returns finding for any address for which alerts in 4 stages were observed in a given time window
+    :return: findings: list
+    """
+    global ALERTED_ADDRESSES
+    global MUTEX
+
+    if not MUTEX:
+        MUTEX = True
+
+        # get time for block to derive date range for query
+        end_date = datetime.utcfromtimestamp(block_event.block.timestamp)
+        start_date = end_date - timedelta(days=DATE_LOOKBACK_WINDOW_IN_DAYS)
+        logging.info(f"Analyzing alerts from {start_date} to {end_date}")
+
+        # get all alerts for date range
+        df_forta_alerts = forta_explorer.empty_alerts()
+        for agent_id in AGENT_IDS:
+            agent_alerts = forta_explorer.alerts_by_agent(agent_id, start_date, end_date)
+            df_forta_alerts = pd.concat([df_forta_alerts, agent_alerts])
+            logging.info(f"Fetched {len(agent_alerts)} for agent {agent_id}")
+
+        # get all addresses that were part of the alerts
+        # to optimize, we only check money laundering addresses as this is required to fullfill all 4 stage requirements
+        money_laundering = df_forta_alerts[df_forta_alerts["alertId"] == "POSSIBLE-MONEY-LAUNDERING-TORNADO-CASH"]
+
+        addresses = set()
+        for index, row in money_laundering.iterrows():
+            addresses = addresses.union(set(row['addresses']))
+
+        # analyze each address' alerts
+        for potential_attacker_address in addresses:
+            logging.debug(potential_attacker_address)
+            # if address is a contract or null address, skip
+            if(is_contract(w3, potential_attacker_address) or potential_attacker_address[0:12] == '0x0000000000'):
+                continue
+
+            # map each alert to 4 stages
+            stages = set()
+            involved_addresses = set()
+            address_alerts = df_forta_alerts[df_forta_alerts["addresses"].apply(lambda x: potential_attacker_address in x)]
+            involved_alert_ids = address_alerts["alertId"].unique()
+            for alert_id in involved_alert_ids:
+                if alert_id in ALERT_ID_STAGE_MAPPING.keys():
+                    stage = ALERT_ID_STAGE_MAPPING[alert_id]
+                    stages.add(stage)
+                    # get addresses from address field to add to involved_addresses
+                    address_alerts[address_alerts["alertId"] == alert_id]["addresses"].apply(lambda x: involved_addresses.update(set(x)))
```

Comment:
`apply` will return a new dataframe, and there's no assignment here, so the updated dataframe is not saved.

Reply:
Mariko, it is as the update operates on the involved_addresses set.